typescript/web/src/components/layout/top-bar/user-menu.tsx

```diff
@@ -57,162 +60,207 @@ const userQuery = gql`
+      borderRadius="full"
+      aria-label="User and Preferences"
+      icon={<UserMenuButtonIcon {...props} />}
+      variant="ghost"
+    />
+  </Tooltip>
+);
+
+const SignOutMenuItem = (props: MenuItemProps) => {
   const removeLastVisitedWorkspaceSlugCookie = useCookies([
     "lastVisitedWorkspaceSlug",
   ])[2];
+  const handleClick = useCallback(() => {
+    removeLastVisitedWorkspaceSlugCookie("lastVisitedWorkspaceSlug", {
+      path: "/",
+      httpOnly: false,
+    });
+    trackEvent("signout", {});
+    signOut({ callbackUrl: "/" });
+  }, [removeLastVisitedWorkspaceSlugCookie]);
+  return (
+    <MenuItem
+      icon={<SignoutIcon fontSize="lg" />}
+      onClick={handleClick}
+      {...props}
+    >
+      Sign out
+    </MenuItem>
+  );
+};
 
-  const userInfoFromSession = session?.user;
+const SignInMenuItem = () => {
+  const [, setIsSigninOpen] = useQueryParam("modal-signin", BoolParam);
+  return (
+    <MenuItem
+      icon={<SigninIcon fontSize="lg" />}
+      onClick={() => setIsSigninOpen(true, "replaceIn")}
+    >
+      Sign in
+    </MenuItem>
+  );
+};
 
-  const { data: userData } = useQuery(userQuery, {
-    variables: { id: userInfoFromSession?.id },
-    skip: userInfoFromSession?.id == null,
-  });
+const SettingsMenuItem = () => {
+  const router = useRouter();
+  const handleClick = useCallback(
+    () => router.push("/settings/profile"),
+    [router]
+  );
+  return (
+    <MenuItem icon={<SettingsIcon fontSize="lg" />} onClick={handleClick}>
+      Settings
+    </MenuItem>
+  );
+};
 
-  const user = userData?.user;
+const UserMenuGroup = () => {
+  const { status } = useSession({ required: false });
+  return (
+    <MenuGroup title="User">
+      {status === "loading" && <SignOutMenuItem cursor="default" disabled />}
+      {status === "authenticated" && (
+        <>
+          <SignOutMenuItem />
+          <SettingsMenuItem />
+        </>
+      )}
+      {status === "unauthenticated" && <SignInMenuItem />}
+    </MenuGroup>
+  );
+};
 
-  const fadeColor = useColorModeValue("gray.600", "gray.400");
-  const avatarBackground = useColorModeValue("white", "gray.700");
+const DarkModeMenuItem = () => {
+  const { toggleColorMode } = useColorMode();
+  return (
+    <MenuItem
+      icon={<DarkModeIcon fontSize="lg" />}
+      closeOnSelect={false}
+      onClick={toggleColorMode}
+    >
+      Switch to dark mode
+    </MenuItem>
+  );
+};
 
-  const router = useRouter();
+const LightModeMenuItem = () => {
+  const { toggleColorMode } = useColorMode();
+  return (
+    <MenuItem
+      icon={<LightModeIcon fontSize="lg" />}
+      closeOnSelect={false}
+      onClick={toggleColorMode}
+    >
+      Switch to light mode
+    </MenuItem>
+  );
+};
 
-  const displayName = user != null ? getDisplayName(user) : "Anonymous";
+const ColorModeMenuItem = () => {
+  const { colorMode } = useColorMode();
+  return colorMode === "light" ? <DarkModeMenuItem /> : <LightModeMenuItem />;
+};
```

Comment:
Why not simplify it to:

```TS
const ColoredColorModeMenuItem = (props: {
  name: string;
  icon: ChakraComponent<IconType, {}>;
}) => {
  const { name, icon: IconComponent } = props;
  const { toggleColorMode } = useColorMode();
  return (
    <MenuItem
      icon={<IconComponent fontSize="lg" />}
      closeOnSelect={false}
      onClick={toggleColorMode}
    >
      Switch to {name} mode
    </MenuItem>
  );
};

const ColorModeMenuItem = () => {
  const { colorMode } = useColorMode();
  return (
    <ColoredColorModeMenuItem
      name={colorMode}
      icon={colorMode === "light" ? LightModeIcon : DarkModeIcon}
    />
  );
};
```
Or even make one function only?

Reply:
Because I've done the refactor semantically but didn't took the 80% time of Pareto's law to finish the last 20% of the splitting-job 😉 
I'll do the change. Thanks!